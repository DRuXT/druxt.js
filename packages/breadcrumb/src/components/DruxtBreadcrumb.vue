<script>
import DruxtModule from 'druxt/dist/components/DruxtModule.vue'
import { mapActions, mapState } from 'vuex'

/**
 * The `<DruxtBreadcrumb />` component uses the DruxtRouter module to build a
 * path based breadcrumb.
 *
 * @example @lang vue
 * <DruxtBreadcrumb />
 */
export default {
  name: 'DruxtBreadcrumb',

  extends: DruxtModule,

  /** */
  props: {
    /**
     * Show home crumb?
     *
     * @type {boolean}
     * @default true
     * @example @lang vue
     * <DruxtBreacrumb :home="false" />
     */
    home: {
      type: Boolean,
      default: true
    }
  },

  /**
   * The Nuxt Fetch hook.
   *
   * Fetches the breadcrumbs.
   */
  async fetch() {
    if (!this.value) {
      await this.fetchCrumbs()
    }
    await DruxtModule.fetch.call(this)
  },

  /** */
  computed: {
    /**
     * @property {objects[]} crumbs - The Breadcrumbs.
     */
    crumbs: ({ model }) => model,

    ...mapState({
      route: state => state.druxtRouter.route,
      routes: state => state.druxtRouter.routes
    })
  },

  /** */
  watch: {
    /**
     * Updates crumbs on Route change.
     */
    $route: async function() {
      await this.$fetch()
    }
  },

  methods: {
    /**
     * Fetch Crumbs
     */
    async fetchCrumbs() {
      // If there is no route, stop here.
      if (!this.route || !Object.keys(this.route).length) return

      // If we are at the root and don't want a home crumb, stop here.
      if (this.$route.path === '/' && !this.home) return

      // Current route crumb.
      const crumbs = []
      if (this.route.label) {
        crumbs.push({ text: this.route.label })
      }

      // If we are at the root of the site, stop here.
      if (this.$route.path === '/') {
        this.model = crumbs
        return
      }

      // Add crumbs for route parents.
      const paths = this.$route.path.split('/').filter(String)
      paths.pop()
      while (paths.length > 0) {
        const to = '/' + paths.join('/')

        let route
        try {
          route = await this.getRoute(to)
        } catch(err) {
          route = false
        }

        if (route.label) {
          crumbs.push({ to, text: route.label })
        }

        paths.pop()
      }

      // Home crumb.
      if (this.home) {
        crumbs.push({
          to: '/',
          text: 'Home'
        })
      }

      this.model = crumbs.reverse()
    },

    /**
     * Maps `druxtRouter/getRoute` Vuex action to `this.getRoute`.
     */
    ...mapActions({
      getRoute: 'druxtRouter/getRoute'
    })
  },

  /** DruxtModule settings */
  druxt: {
    /**
     * Provides the available component naming options for the Druxt Wrapper.
     *
     * @param {object} context - The module component ViewModel.
     * @returns {ComponentOptions}
     */
    componentOptions: ({}) => [['default']],

    /**
     * Provides propsData for the DruxtWrapper.
     *
     * @param {object} context - The module component ViewModel.
     * @returns {PropsData}
     */
    propsData: ({ model }) => ({ crumbs: model, value: model }),

    /**
     * Provides the scoped slots object for the Module render function.
     *
     * The `default` slot renders crumbs as a list of NuxtLink's.
     *
     * @return {ScopedSlots} The Scoped slots object.
     */
    slots() {
      // Build scoped slots for each field.
      const scopedSlots = {}

      // Build default slot.
      scopedSlots.default = () => this.$createElement('ul', this.crumbs.map((crumb) =>
        this.$createElement('li', [
          crumb.to
            ? this.$createElement('NuxtLink', { props: { to: crumb.to }}, [crumb.text])
            : crumb.text
        ])
      ))

      return scopedSlots
    },
  },
}

/**
 * Provides the available naming options for the Wrapper component.
 *
 * @typedef {array[]} ComponentOptions
 *
 * @example @lang js
 * [
 *   'DruxtBreadcrumb[Default]',
 * ]
 *
 * @example @lang js
 * [
 *   'DruxtBreadcrumbDefault',
 * ]
 */

/**
 * Provides propsData for use in the Wrapper component.
 *
 * @typedef {object} PropsData
 * @param {objects[]} crumbs - The Breadcrumbs.
 * @param {objects[]} value - The Breadcrumbs value.
 *
 * @example @lang js
 * {
 *   crumbs: [{
 *     text: 'Home',
 *     to: '/',
 *   }],
 *   value: [{
 *     text: 'Home',
 *     to: '/',
 *   }],
 * }
 */

/**
 * Provides scoped slots for use in the Wrapper component.
 *
 * @typedef {object} ScopedSlots
 * @param {function} default - Crumbs as a list of NuxtLink's.
 *
 * @example <caption>DruxtBreadcrumb**Default**.vue</caption> @lang vue
 * <template>
 *   <div>
 *     <slot />
 *   </div>
 * </template>
 */
</script>
