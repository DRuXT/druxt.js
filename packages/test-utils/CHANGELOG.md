# druxt-test-utils

## 0.2.15

### Patch Changes

- Updated dependencies [41cab3a0]
  - druxt@0.22.0

## 0.2.14

### Patch Changes

- Updated dependencies [e46a3290]
  - druxt@0.21.0

## 0.2.13

### Patch Changes

- Updated dependencies [49b67872]
  - druxt@0.20.0

## 0.2.12

### Patch Changes

- Updated dependencies [be21952]
  - druxt@0.19.0

## 0.2.11

### Patch Changes

- Updated dependencies [e6cf1fc]
  - druxt@0.18.3

## 0.2.10

### Patch Changes

- 540afca: Updated drupal-jsonapi-params to 2.0.0
- Updated dependencies [540afca]
  - druxt@0.18.2

## 0.2.9

### Patch Changes

- Updated dependencies [80164a1]
  - druxt@0.18.1

## 0.2.8

### Patch Changes

- Updated dependencies [e3d5238]
  - druxt@0.18.0

## 0.2.7

### Patch Changes

- Updated dependencies [b79701c]
  - druxt@0.17.0

## 0.2.6

### Patch Changes

- Updated dependencies [7b749bd]
  - druxt@0.16.0

## 0.2.5

### Patch Changes

- Updated dependencies [2ae1d6d]
  - druxt@0.15.0

## 0.2.4

### Patch Changes

- Updated dependencies [45bc0b9]
  - druxt@0.14.0

## 0.2.3

### Patch Changes

- Updated dependencies [77ab204]
  - druxt@0.13.0

## 0.2.2

### Patch Changes

- Updated dependencies [4504a2f]
  - druxt@0.12.0

## 0.2.1

### Patch Changes

- Updated dependencies [75ff8a9]
  - druxt@0.11.0

## 0.2.0

### Patch Changes

- Updated dependencies [715e5ef]
  - druxt@0.10.0

## 0.1.0

### Patch Changes

- Updated dependencies [21170fb]
  - druxt@0.9.0
