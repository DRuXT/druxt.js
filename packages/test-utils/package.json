{
  "name": "druxt-test-utils",
  "version": "0.1.0",
  "description": "Test utilities for Druxt development.",
  "keywords": [
    "druxt",
    "jsonapi",
    "test"
  ],
  "exports": {
    ".": {
      "require": "./dist/druxt-test-utils.ssr.js",
      "import": "./dist/druxt-test-utils.esm.js"
    }
  },
  "main": "dist/druxt-test-utils.ssr.js",
  "module": "dist/druxt-test-utils.esm.js",
  "dependencies": {
    "drupal-jsonapi-params": "^1.2.2",
    "jest-mock-axios": "^4.4.0"
  },
  "peerDependencies": {
    "druxt": "*"
  }
}
