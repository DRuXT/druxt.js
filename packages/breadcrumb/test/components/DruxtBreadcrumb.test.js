import 'regenerator-runtime/runtime'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import { DruxtRouterStore } from 'druxt-router'
import Vuex from 'vuex'

import DruxtBreadcrumb from '../../src/components/DruxtBreadcrumb.vue'

// Setup local vue instance.
const localVue = createLocalVue()
localVue.use(Vuex)

let store

const mountComponent = ({ path, routes, propsData, options }) => {
  // Setup mocks.
  const mocks = {
    $route: { path },
    $fetchState: { pending: false }
  }

  if (typeof routes === 'undefined') {
    routes = [path]
  }

  // Add mock route to store.
  for (const route of routes) {
    store.commit('druxtRouter/addRoute', {
      path: route,
      route: {
        label: route
      }
    })
  }
  if (typeof store.state.druxtRouter.routes[path] !== 'undefined') {
    store.commit('druxtRouter/setRoute', path)
  }

  return shallowMount(DruxtBreadcrumb, { store, localVue, mocks, propsData, ...options })
}

describe('DruxtBreadcrumb', () => {
  beforeEach(() => {
    // Setup vuex stores.
    store = new Vuex.Store()
    store.$druxtRouter = jest.fn(() => ({
      getRoute: async (to) => {
        if (to === '/error') throw new Error('Error')
        return { data: {} }
      }
    }))
    DruxtRouterStore({ store })
  })

  test('root', async () => {
    const wrapper = mountComponent({ path: '/' })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    expect(wrapper.vm.route).toStrictEqual({
      label: '/'
    })
    expect(Object.keys(wrapper.vm.routes)).toHaveLength(1)

    expect(wrapper.vm.crumbs).toHaveLength(1)
    expect(wrapper.vm.crumbs[0].text).toBe('/')

    expect(wrapper.html()).toMatchSnapshot()

    // Default slot.
    const slot = wrapper.vm.getScopedSlots().default()
    expect(slot.tag).toBe('ul')
    expect(slot.children[0].tag).toBe('li')
  })

  test('level 1', async () => {
    const wrapper = mountComponent({ path: '/level-1', routes: ['/', '/level-1'] })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    expect(wrapper.vm.route).toStrictEqual({
      label: '/level-1'
    })

    expect(wrapper.vm.crumbs).toHaveLength(2)
    expect(wrapper.vm.crumbs[0].to).toBe('/')

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('level 2', async () => {
    const wrapper = mountComponent({ path: '/level-1/level-2', routes: ['/', '/level-1', '/level-1/level-2'] })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    expect(wrapper.vm.route).toStrictEqual({
      label: '/level-1/level-2'
    })

    expect(wrapper.vm.crumbs).toHaveLength(3)
    expect(wrapper.vm.crumbs[0].to).toBe('/')

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('missing parent', async () => {
    const wrapper = mountComponent({ path: '/level-1/level-2', routes: ['/', '/level-1/level-2'] })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    expect(wrapper.vm.route).toStrictEqual({
      label: '/level-1/level-2'
    })

    expect(wrapper.vm.crumbs).toHaveLength(2)
    expect(wrapper.vm.crumbs[0].to).toBe('/')

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('404 route', async () => {
    const wrapper = mountComponent({ path: '/level-1/level-2', routes: ['/'] })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    expect(wrapper.vm.route).toStrictEqual({})
    expect(wrapper.vm.crumbs).toBe(null)

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('error', async () => {
    const wrapper = mountComponent({ path: '/error/level-2', routes: ['/', '/error/level-2'] })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    expect(wrapper.vm.route).toStrictEqual({
      label: '/error/level-2'
    })

    expect(wrapper.vm.crumbs).toHaveLength(2)
    expect(wrapper.vm.crumbs[0].to).toBe('/')

    // @TODO - Inconsistent behaviour between local and CI. Investigate.
    // expect(wrapper.html()).toMatchSnapshot()
  })

  test('no home', async () => {
    const wrapper = mountComponent({ path: '/', propsData: { home: false } })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    expect(wrapper.vm.crumbs).toBe(null)

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('custom default slot', async () => {
    const scopedSlots = { default: jest.fn() }
    const wrapper = mountComponent({ path: '/', options: { scopedSlots } })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    wrapper.vm.getScopedSlots().default.call()
    expect(scopedSlots.default).toHaveBeenCalledWith({
      crumbs: [{
        text: '/',
      }],
      value: [{
        text: '/',
      }],
    })
  })
})
