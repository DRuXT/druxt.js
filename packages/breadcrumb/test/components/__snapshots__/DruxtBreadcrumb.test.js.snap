// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DruxtBreadcrumb 404 route 1`] = `
"<div>
  <druxtwrapper-stub></druxtwrapper-stub>
</div>"
`;

exports[`DruxtBreadcrumb level 1 1`] = `
"<div>
  <druxtwrapper-stub crumbs=\\"[object Object],[object Object]\\" value=\\"[object Object],[object Object]\\"></druxtwrapper-stub>
</div>"
`;

exports[`DruxtBreadcrumb level 2 1`] = `
"<div>
  <druxtwrapper-stub crumbs=\\"[object Object],[object Object],[object Object]\\" value=\\"[object Object],[object Object],[object Object]\\"></druxtwrapper-stub>
</div>"
`;

exports[`DruxtBreadcrumb missing parent 1`] = `
"<div>
  <druxtwrapper-stub crumbs=\\"[object Object],[object Object]\\" value=\\"[object Object],[object Object]\\"></druxtwrapper-stub>
</div>"
`;

exports[`DruxtBreadcrumb no home 1`] = `
"<div>
  <druxtwrapper-stub></druxtwrapper-stub>
</div>"
`;

exports[`DruxtBreadcrumb root 1`] = `
"<div>
  <druxtwrapper-stub crumbs=\\"[object Object]\\" value=\\"[object Object]\\"></druxtwrapper-stub>
</div>"
`;
